### PR TITLE
Fixing syntax error reporting in the presence of token lookahead

### PR DESCRIPTION
**Pull request by herbelin**
**Kind**: between fix and enhancement

A first commit changes the syntax error message in the following case:
```coq
Notation "## x" := (S x) (at level 0).
Notation "## ##" := 0 (at level 0).
Check ## [].
```
Before: `Syntax error: [term] expected after '##' (in [term])`.
After: `Syntax error: '##' or [term] expected after '##' (in [term])`.

A second commit refines the situation further so that it works also in the presence of several tokens ahead:
```coq
Notation "## x" := (S x) (at level 0).
Notation "## ## ##" := 0 (at level 0).
Check ## [].
```
Before: `Syntax error: [term] expected after '##' (in [term])`.
With only the first commit: `Syntax error: '##' '##' expected after '##' (in [term]).`
After: `Syntax error: '##' '##' or [term] expected after '##' (in [term]).`

The first problem was that the error was reported on the last applicable rule rather than on the list of all applicable rules at this stage. The second problem was that in the presence of multiple tokens ahead, the alternative rules were omitted.

***Note: I rewrote the header in a more concise and comprehensive way, compared to the initial version, also fixing the initial PR and adding a new fix.***

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by herbelin** [comment]
Even if a priori w/o effects since the CI tests only syntactically correct code:
@coqbot run full ci

**Comment by herbelin** [comment]
This seems ready from the PR point of view. @SkySkimmer, maybe would you like to assign?

**Comment by Alizter** [comment]
@coqbot bench

File: gramlib/grammar.ml [hunk 1]
```diff
@@ -1293,8 +1297,12 @@ and parser_of_token_list : type s tr lt r.
...
            (match try Some (p1 gstate a strm) with Stream.Failure -> None with
             | Some act -> act a
-            | None -> p2 gstate last_a strm)
-         | None -> p2 gstate last_a strm)
+            | None ->
+              (try p2 gstate last_a strm
+               with TokenListFailed _ -> raise (TokenListFailed (entry, a, Stoken tok, son))))
+         | None ->
+            (try p2 gstate last_a strm
+             with TokenListFailed _ -> raise (TokenListFailed (entry, last_a, Stoken (last_tok), tree))))
```
Comment (SkySkimmer) [inline gramlib/grammar.ml:1305]:
```suggestion
             with TokenListFailed _ -> raise (TokenListFailed (entry, last_a, Stoken last_tok, tree))))
```

**Review by SkySkimmer — APPROVED** [review]
How about adding some output test?
If you put it in output-coqtop/ you can test mutliple errors in 1 file

**Comment by SkySkimmer** [comment]
ping @herbelin do you want to add a test?

File: gramlib/grammar.ml (hunk 1 above)
Comment (herbelin) [reply gramlib/grammar.ml]: Done

**Comment by herbelin** [comment]
> ping @herbelin do you want to add a test?

Done

**Comment by SkySkimmer** [comment]
@coqbot run full ci

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**